libresin/libresin/core/resources/shader_resource.cpp

```diff
@@ -0,0 +1,202 @@
+#include <cctype>
+#include <filesystem>
+#include <fstream>
+#include <libresin/core/resources/shader_resource.hpp>
+#include <libresin/helpers/logger_helpers.hpp>
+#include <libresin/utils/exceptions.hpp>
+#include <libresin/utils/logger.hpp>
+#include <optional>
+#include <ranges>
+#include <string_view>
+#include <unordered_set>
+
+namespace resin {
+ShaderResource::ShaderResource(std::string&& content, ShaderType type, std::unordered_set<std::string>&& ext_defi_names)
+    : ext_defi_names_(std::move(ext_defi_names)), raw_content_(std::move(content)), type_(type), is_dirty_(true) {}
+
+const std::unordered_set<std::string>& ShaderResource::get_ext_defi_names() const { return ext_defi_names_; }
+
+void ShaderResource::set_ext_defi(std::string_view ext_defi_name, std::string&& defi_content) {
+  auto it = std::ranges::find(ext_defi_names_, ext_defi_name);
+  if (it == ext_defi_names_.end()) {
+    resin::Logger::warn("Shader resource could not find external definition named \"{}\"", ext_defi_name);
+    return;
+  }
+
+  ext_defi_contents_[*it] = std::move(defi_content);
+  is_dirty_               = true;
+}
+
+bool ShaderResource::is_glsl_ready() const { return ext_defi_contents_.size() == ext_defi_names_.size(); }
+
+const std::string& ShaderResource::get_raw() const { return raw_content_; }
+
+const std::string& ShaderResource::get_glsl() const {
+  if (!is_dirty_) {
+    return glsl_;
+  }
+
+  glsl_.clear();
+  glsl_.append(shader_macros::kSupportedShaderVersion);
+  for (const auto& ext_defi : ext_defi_contents_) {
+    glsl_.append(std::format("#define {} {}\n", ext_defi.first, ext_defi.second));
+  }
+  glsl_.append(raw_content_);
+
+  is_dirty_ = false;
+  return glsl_;
+}
+
+static ShaderType get_sh_type(const std::filesystem::path& path) {
+  auto file_ext = path.extension().string();
+  auto sh_type  = extension_to_shader_type(file_ext);
+  if (!sh_type.has_value()) {
+    helpers::log_file_err(path, "File extension \"{}\" is not suported.", file_ext);
+
+    throw FileExtensionNotSupportedException();
+  }
+
+  return sh_type.value();
+}
+
+static std::string load_content(const std::filesystem::path& path) {
+  namespace fs = std::filesystem;
+
+  if (!fs::exists(path)) {
+    helpers::log_file_err(path, "File does not exist.");
+    throw FileDoesNotExistException();
+  }
+
+  if (!fs::is_regular_file(path) && !fs::is_symlink(path)) {
+    helpers::log_file_err(path, "File must be a regular file or symlink.");
+    throw InvalidFileTypeException();
+  }
+
+  std::ifstream file_stream(path.string());
+  if (!file_stream.is_open()) {
+    helpers::log_file_err(path, "File stream cannot be opened.");
+    throw FileStreamNotAvailableException();
+  }
+
+  std::stringstream buffer;
+  buffer << file_stream.rdbuf();
+  return buffer.str();
+}
+
+static std::optional<std::filesystem::path> process_include_macro(std::string_view arg, int64_t curr_line_num) {
```

Comment:
```suggestion
static std::optional<std::filesystem::path> process_include_macro_arg(std::string_view arg, int64_t curr_line_num) {
```
and `process_include_macro` method handling whole macro